finality-verifiers/grandpa/src/bridges/runtime/storage_proof.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Logic for checking Substrate storage proofs.
+
+use hash_db::{HashDB, Hasher, EMPTY_PREFIX};
+use sp_runtime::RuntimeDebug;
+use sp_std::vec::Vec;
+use sp_trie::{read_trie_value, LayoutV1, MemoryDB, StorageProof};
+
+/// This struct is used to read storage values from a subset of a Merklized database. The "proof"
+/// is a subset of the nodes in the Merkle structure of the database, so that it provides
+/// authentication against a known Merkle root as well as the values in the database themselves.
+pub struct StorageProofChecker<H>
+where
+    H: Hasher,
+{
+    root: H::Out,
+    db: MemoryDB<H>,
+}
+
+impl<H> StorageProofChecker<H>
+where
+    H: Hasher,
+{
+    /// Constructs a new storage proof checker.
+    ///
+    /// This returns an error if the given proof is invalid with respect to the given root.
+    pub fn new(root: H::Out, proof: StorageProof) -> Result<Self, Error> {
+        let db = proof.into_memory_db();
+        if !db.contains(&root, EMPTY_PREFIX) {
+            return Err(Error::StorageRootMismatch)
+        }
+
+        let checker = StorageProofChecker { root, db };
+        Ok(checker)
+    }
+
+    /// Reads a value from the available subset of storage. If the value cannot be read due to an
+    /// incomplete or otherwise invalid proof, this returns an error.
+    pub fn read_value(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Error> {
+        // LayoutV1 or LayoutV0 is identical for proof that only read values.
+        read_trie_value::<LayoutV1<H>, _>(&self.db, &self.root, key)
+            .map_err(|_| Error::StorageValueUnavailable)
+    }
+}
+
+#[derive(RuntimeDebug, PartialEq)]
+pub enum Error {
+    StorageRootMismatch,
+    StorageValueUnavailable,
+}
+
+/// Return valid storage proof and state root.
+///
+/// NOTE: This should only be used for **testing**.
```

Comment:
```suggestion
/// NOTE: This should only be used for **testing**.
#[cfg(test)]
```

Reply:
I've solved this via feature flag, as this is how I handle it in the entire pallet